src/Tooltip/Tooltip.js

```diff
@@ -0,0 +1,141 @@
+import classnames from 'classnames';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+const METHODS_TO_BIND = [
+  'getIdealPosition',
+  'handleMouseEnter',
+  'handleMouseLeave'
+];
+
+class Tooltip extends React.Component {
+  constructor() {
+    super();
+
+    this.state = {isOpen: false};
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  getIdealPosition() {
+    // Get the anchor and position from state if possible. If not, get it from
+    // the props.
+    let anchor = this.state.anchor || this.props.anchor;
+    let position = this.state.position || this.props.position;
+
+    if (this.refs.tooltipContent) {
+      let viewportHeight = global.window.innerHeight;
+      let viewportWidth = global.window.innerWidth;
+      let tooltipContent = ReactDOM.findDOMNode(this.refs.tooltipContent);
+      let tooltipPosition = tooltipContent.getBoundingClientRect();
+
+      // Change the position if the tooltip will be rendered off the screen.
+      if (position === 'right' && tooltipPosition.right > viewportWidth) {
+        position = 'left';
+      } else if (position === 'left' && tooltipPosition.left < 0) {
+        position = 'right';
+      } else if (position === 'top' && tooltipPosition.top < 0) {
+        position = 'bottom';
+      } else if (position === 'bottom' &&
+        tooltipPosition.bottom > viewportHeight) {
+        position = 'top';
+      }
+
+      // Change the anchor if the tooltip will be rendered off the screen.
+      if (position === 'right' || position === 'left') {
+        if (tooltipPosition.top < 0) {
+          anchor = 'start';
+        } else if (tooltipPosition.bottom > viewportHeight) {
+          anchor = 'end';
```

Comment:
Could we default the anchor to be 'end' or 'start'? I believe then you can remove the nested if-statements and just have a if-else.